src/board/devices/output/lcd.ts

```diff
@@ -0,0 +1,252 @@
+    }
+    switch (this.lcdPositionInfo[position]) {
+      case LcdPositionType.NotSet: {
+        return ''
+      }
+      case LcdPositionType.Ascii: {
+        let val = this.memory.readByte(
+          LcdDisplay.LCD_ASCII_ADRESS_LIST[position]
+        )
+        if (val.value == 0) {
+          return ''
+        } else {
+          return String.fromCharCode(val.toUnsignedInteger())
+        }
+      }
+      case LcdPositionType.Binary: {
+        // Check if this position is an empty field (offset = -1)
+        if (LcdDisplay.LCD_BINARY_FIELD_CONFIGURATION[position][0] == -1) {
+          return ''
+        }
+        let binaryString: string = this.memory
+          .readByte(
+            LcdDisplay.LCD_BINARY_ADRESS_LIST[
+              LcdDisplay.LCD_BINARY_FIELD_CONFIGURATION[position][0]
+            ]
+          )
+          .toBinaryString()
+        if (parseInt(binaryString) == 0) {
+          return ''
+        }
+        // Check if position is the first half of the byte
+        let s: string
+        if (LcdDisplay.LCD_BINARY_FIELD_CONFIGURATION[position][1] == 1) {
+          s = binaryString.substring(0, 4)
+        } else {
+          s = binaryString.substring(4, 8)
+        }
+        return parseInt(s, 2).toString(16).toUpperCase()
+      }
+    }
+  }
+
+  /**
+   * Returns the actual set colour brightness of the lcd screen
+   *
+   * @returns an array of halfwords with the colour brightness values in the order red, green, blue
+   */
+  public getColour(): Halfword[] {
+    let colour: Halfword[] = []
+    for (let i = 0; i < LcdDisplay.LCD_COLOUR_ADRESS_LIST.length; i += 2) {
+      colour[i / 2] = this.memory.readHalfword(
+        LcdDisplay.LCD_COLOUR_ADRESS_LIST[i]
+      )
+    }
+    return colour
+  }
+
+  /**
+   * Returns the actual set colour brightness of the lcd screen in hex value
+   *
+   * @returns an string of hex color code to use in css
+   */
+  public getColourHex(): string {
+    let colourHex = '#'
+    const colours = this.getColour()
+    for (let col of colours) {
+      colourHex += col.toHexString().substring(2)
+    }
+    return colourHex
+  }
+
+  public writeByte(address: Word, byte: Byte): void {
+    super.writeByte(address, byte)
+    this.writeAtPosition(address)
+  }
+
+  public writeBytes(address: Word, bytes: Byte[]): void {
+    super.writeBytes(address, bytes)
+    for (let i = 0; i < bytes.length; i++) {
+      this.writeAtPosition(address.add(i))
+    }
+  }
+
+  public writeHalfword(address: Word, halfword: Halfword): void {
+    super.writeHalfword(address, halfword)
+    this.writeAtPosition(address)
+    this.writeAtPosition(address.add(1))
+  }
+
+  public writeWord(address: Word, word: Word): void {
+    super.writeWord(address, word)
+    for (let i = 0; i < 4; i++) {
+      this.writeAtPosition(address.add(i))
+    }
+  }
+
+  private writeAtPosition(address: Word): void {
+    for (let i = 0; i < LcdDisplay.LCD_ASCII_ADRESS_LIST.length; i++) {
```

Comment:
Instead of looping you could also find the index mathematically like so (I'm not 100% sure if this code is correct though):
```js
let f = (address, base) => (Math.ceil((address / base)) - 1) * (address  % 0x10)
f(0x60000330, 0x60000330) // prints 0
f(0x60000331, 0x60000330) // prints 1
f(0x60000332, 0x60000330) // prints 2
f(0x60000333, 0x60000330) // prints 3
....
```

Reply:
@stefenz this one is for you

Reply:
Even easier is to just subtract the base from the address. But the problem is, that i don't know if the address is in the ascii or in the binary range.

I tried to improve it a little bit. I'm curious about your feedback!

Reply:
But couldn't you just simply check in which range the incoming address is? Because you have your Word arrays with all the addresses.

Reply:
Something like this for ASCII:
```js
if (address >= LCD_ASCII_ADRESS_LIST[0] || address <= LCD_ASCII_ADRESS_LIST[LCD_ASCII_ADRESS_LIST.length-1]) {
   // ...
}
```

Reply:
But if it gets too complicated we can also leave it as it is. I mean the code works its just making the code more beautiful.

Reply:
> ```js
> ] ||
> ```

I would connect with AND (&&)

Reply:
Did you mean like that?

Reply:
Yeah thats how I imagined it!